gnome-image-installer/util/gis-unattended-config.c

```diff
@@ -409,3 +409,109 @@ gis_unattended_config_match_computer (GisUnattendedConfig *self,
+    }
+
+  return g_file_get_child (dir, target_basename);
+}
+
+static gboolean
+rename_existing_file_if_exists (GFile   *source,
+                                gchar  **backup_path,
+                                GError **error)
+{
+  g_autoptr(GFile) target = invent_backup_name (source);
+  g_autoptr(GError) local_error = NULL;
+
+  if (g_file_move (source, target,
+                   /* No G_FILE_COPY_OVERWRITE, so target will not be
+                    * overwritten if it already exists. Since its name includes
+                    * a fresh UUID, this is extremely unlikely.
+                    */
+                   G_FILE_COPY_NO_FALLBACK_FOR_MOVE,
+                   NULL, NULL, NULL, &local_error))
+    {
+      *backup_path = g_file_get_basename (target);
+      return TRUE;
+    }
+
+  if (g_error_matches (local_error,
+                       G_IO_ERROR,
+                       G_IO_ERROR_NOT_FOUND))
+    {
+      backup_path = NULL;
+      return TRUE;
+    }
+
+  g_propagate_error (error, g_steal_pointer (&local_error));
+  return FALSE;
+}
+
+/**
+ * gis_unattended_config_write:
+ * @target_path: where to write new unattended config. If it already exists,
+ *  the existing file will be renamed to a unique name in the same directory.
+ * @backup_path: (out): set to the basename of backup of @target_path, if it
+ *  already existed, or to %NULL if it did not.
+ *
+ * Either @vendor and @product must both be %NULL, or they must both be
+ * non-%NULL.
+ */
+gboolean
+gis_unattended_config_write (const gchar *target_path,
+                             const gchar *locale,
+                             const gchar *image,
+                             const gchar *blockdevice,
+                             const gchar *vendor,
+                             const gchar *product,
+                             gchar      **backup_path,
+                             GError     **error)
+{
+  g_autoptr(GFile) target_file = g_file_new_for_path (target_path);
+  g_autoptr(GKeyFile) key_file = g_key_file_new ();
+
+  if (locale != NULL)
+    g_key_file_set_string (key_file, EOS_GROUP, LOCALE_KEY, locale);
+
+  if (image != NULL)
+    g_key_file_set_string (key_file, IMAGE_GROUP_PREFIX, FILENAME_KEY, image);
+
+  if (blockdevice != NULL)
+    g_key_file_set_string (key_file, IMAGE_GROUP_PREFIX, BLOCKDEVICE_KEY,
+                           blockdevice);
+
+  if (vendor != NULL)
+    {
+      g_return_val_if_fail (product != NULL, FALSE);
+      g_key_file_set_string (key_file, COMPUTER_GROUP_PREFIX, VENDOR_KEY,
+                             vendor);
+      g_key_file_set_string (key_file, COMPUTER_GROUP_PREFIX, PRODUCT_KEY,
+                             product);
+    }
+  else
+    {
+      g_return_val_if_fail (product == NULL, FALSE);
+    }
+
+  if (!rename_existing_file_if_exists (target_file, backup_path, error))
+    return FALSE;
+
+  return g_key_file_save_to_file (key_file, target_path, error);
```

Comment:
/me muses about the possibility of using `renameat2()` here to atomically exchange the temp file that the key file is saved to, and the (possibly already existing) `target_path`.

Reply:
(This is not a serious suggestion.)

Reply:
I took this suggestion.

Reply:
:ohno: